crypto/src/threshold/threshold.rs

```diff
@@ -0,0 +1,387 @@
+        let res = Fr::from_le_bytes_mod_order(res_hash.as_bytes());
+
+        if res != proof.t {
+            return Err(anyhow::anyhow!("invalid encryption proof"));
+        }
+        Ok(())
+    }
+
+    /// Add this [`EncryptedValue`] to another [`EncryptedValue`] to produce a new [`EncryptedValue`].
+    /// NOTE: proofs are not aggregatable, so any aggregations must be verified independently.
+    pub fn add(&self, other: &EncryptedValue) -> EncryptedValue {
+        EncryptedValue {
+            c1: self.c1 + other.c1,
+            c2: self.c2 + other.c2,
+            proof: None,
+        }
+    }
+}
+
+/// Encrypt the given `values` as value*decaf377::basepoint using the elgamal scheme, and compute
+/// an [`EncrytionProof`] of correctness.
+pub fn encrypt_value(value: decaf377::Fr, for_pubkey: decaf377::Element) -> EncryptedValue {
+    let mut rng = rand::thread_rng();
+    let e = Fr::rand(&mut rng);
+    let c1 = e * decaf377::basepoint();
+    let c2 = e * for_pubkey + value * decaf377::basepoint();
+
+    let k1 = Fr::rand(&mut rng);
+    let k2 = Fr::rand(&mut rng);
+    let alpha = decaf377::basepoint() * k1;
+    let gamma = for_pubkey * k1 + decaf377::basepoint() * k2;
+    let challenge_hash = blake2b_simd::Params::default()
+        .personal(b"elgenc")
+        .to_state()
+        .update(&c1.compress().0)
+        .update(&c2.compress().0)
+        .update(&for_pubkey.compress().0)
+        .update(&alpha.compress().0)
+        .update(&gamma.compress().0)
+        .finalize();
+    let t = Fr::from_le_bytes_mod_order(challenge_hash.as_bytes());
+    let r = k1 - e * t;
+    let s = k2 - value * t;
+
+    let proof = EncryptionProof { r, s, t };
+    EncryptedValue {
+        c1,
+        c2,
+        proof: Some(proof),
+    }
+}
+
+/// Decrypt a given [`EncryptedValue`] in the threshold setting using a set of decryption shares.
+pub fn decrypt_value(
+    encrypted_value: &EncryptedValue,
+    decryption_shares: &[DecryptionShare],
+    participant_commitments: &[decaf377::Element],
+) -> Result<decaf377::Element, anyhow::Error> {
+    for (i, share) in decryption_shares.iter().enumerate() {
+        share.verify(encrypted_value.c1, participant_commitments[i])?;
+    }
+
+    let indicies = decryption_shares
+        .iter()
+        .map(|share| share.participant_index)
+        .collect::<Vec<_>>();
+
+    let mut d = decaf377::Element::default();
+    for share in decryption_shares {
+        d = d + share.share * lagrange_coefficient(share.participant_index, &indicies);
+    }
+    Ok(-d + encrypted_value.c2)
+}
+
+/// Sum a set of encrypted values together using the additive homomorphism.
+pub fn aggregate_values(
+    values: &[EncryptedValue],
+    dkg_pubkey: decaf377::Element,
+) -> Result<EncryptedValue, anyhow::Error> {
+    let mut res = values.first().unwrap().clone();
+    res.verify(dkg_pubkey)?;
+    for i in 1..values.len() {
+        values[i].verify(dkg_pubkey)?;
+        res = res.add(&values[i]);
```

Comment:
consider:
```suggestion
    for value in values.iter().skip(1) {
        value.verify(dkg_pubkey)?;
        res = res.add(&value);
```
(this should elide unnecessary bounds checks)